src/merkle_tree/mod.rs

```diff
@@ -0,0 +1,184 @@
+use std::ops::Rem;
+
+use crate::{hash, hash_pair};
+
+pub struct MerkleTree {
+    hashes: Vec<Vec<u8>>,
+    count: usize
+}
+
+impl MerkleTree {
+    pub fn new(elements: &[String]) -> MerkleTree{
+        let mut tree = MerkleTree{hashes: vec![], count:  elements.len()};
+        let leaves = tree.hash_elements(elements);
+        tree.hashes = tree.build_hashes(leaves);
+        tree
+    }
+
+    pub fn root(&self) -> &[u8] {
+        &self.hashes[0]
+    }
+
+    pub fn add(&mut self, elements: &[String]) {
+        let new_leaves = self.hash_elements(&elements);
+        let old_leaves = self.hashes[{self.count-1}..].to_vec();
+        let leaves = [old_leaves, new_leaves].concat();
+        let count =  leaves.len();
+        self.hashes = self.build_hashes(leaves);
+        self.count = count;
+    }
+
+    fn hash_elements(&self, elements: &[String]) -> Vec<Vec<u8>> {
+        elements.iter().map(|e| hash(e.to_string())).collect()
+    }
+
+    fn build_hashes(&self, hashes: Vec<Vec<u8>>) -> Vec<Vec<u8>> {
+        if hashes.len() == 1 {
+            return hashes;
+        }
+
+        let h: Vec<Vec<u8>>= hashes.chunks(2).clone().into_iter().map(|e| hash_pair(e[0].clone(), e[1].clone())).collect();
+        [self.build_hashes(h), hashes].concat()
+    }
+
+    fn proof(&self, mut index: usize) -> Vec<Vec<u8>> {
+        let mut proof = vec![];
+        let mut i = self.count-1;
+
+        while i != 0 {
+            let h: Vec<u8> = match index.rem(2) {
+                0 => self.hashes[index+1+i].clone(),
+                _ => self.hashes[index+i-1].clone()
+            };
+            proof.append(&mut vec![h]);
+            index = index/2;
+            i = {i+1}/2 - 1;
+        }
+        proof
+    }
+
+    fn verify(&self, proof: &[Vec<u8>], mut index: usize) -> bool {
+        let mut hash = self.hashes[self.count-1+index].clone();
+        for p in proof.iter() {
```

Comment:
I believe you don't need a call to `iter` here. The `for` loop does an implicit call to `into_iter` anyways.